elastic/apm-server
# testing/smoke: Add initial smoke tests

## Motivation/summary

This patch adds an initial set of smoke tests which use the existing
`ec_deployment` module for the underlying infrastructure and a bash
script to drive the scripted smoke test cases. The script takes in a
`MAJOR.MINOR` version combination.

The initial smoke tests are slightly different depending if the major
version is `7` or `8`, the former will be test upgrades `7.17.last-1` to
`7.17.last`, while the latter, will upgrade from the previous minor to
the current minor (`8.3.0`) will upgrade from `8.2.3` to `8.3.0`.

The initial set of assertions is kept to a minimum to enable iterating
over the approach rather than adding all the assertions in the first PR.

Initially, we're testing that:

- Create deployment succeeds
- Data is sent to APM Server
- Each of the events sent to APM Server can be found in Elasticsearch.
- Upgrade to the next version succeeds.
- Data is sent to the APM Server
- Each of the events sent to APM Server can be found in Elasticsearch.

As the last step and if any errors occur during the smoke test, the
underlying infrastructure will be destroyed.

I've chosen to use bash since it doesn't seem too bad with the current
set of assertions, we can always improve it and write it in Go, but it
seems more involved than what we need right now.

I've created a [gist with the output](https://gist.github.com/marclop/81d567ecafd23058c221e41f147ee2aa) of the tested execution.

## Checklist

~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~
~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~

## How to test these changes

1. `export your cloud credentials (export EC_API_KEY)`
2. `make smoketest SMOKETEST_VERSIONS=7.17,latest`
3. Verify that the exit status is 0 and that the smoke tests all passed (check for 4 instances of the log message `-> Smoke tests passed!`)

## Related issues

Part of #8303
